Get `use VERSION` to import the corresponding `builtin` bundle

leonerd opened the pull request:
A few months ago we added `use builtin :VER` bundles, specifically with the plan that we'd eventually get `use VERSION` to imply it. This PR finally completes the last step of the process, allowing a simple `use VERSION` declaration to import all of the (stable) builtin symbols at that version.

Now the full user convenience is

```perl
use v5.40;
say "True is ", true;
weaken( my $weakref = [] );
# etc...
```

This also applies to the `-E` commandline switch, allowing short oneliners like

```
$ perl -E 'say reftype []'
ARRAY
```

Grinnz commented on pod/perlfunc.pod:
```diff
@@ -10177,6 +10183,21 @@ file (possibly after a C<package> statement or any amount of whitespace or
+This is because any imported functions from C<builtin> would need to be made
+invisible again.  If you wish to use a higher version only in a small
+section of a file, enclose that section in its own lexical scope.  Once
+outside the scope again afterwards, any imported effects are no longer
```
perhaps "any lexically imported effects" or "any imported effects from use VERSION" so as not to be misleading

leonerd replied:
Ahyes; reworded in 3fadae8c81 (will fixup before merge)

happy-barney commented on t/comp/use.t:
```diff
@@ -157,6 +157,12 @@ like $@, qr/^Can't use string/,
+    'downgrading prevailing version in same scope is not permitted';
+eval '{ use 5.039; } use 5.020;';
+ok !$@, 'downgrading prevailing version by falling out of scope is permitted';
+
```
shouldn't this test also availability of symbols ?

happy-barney replied:
also it would be nice to test behaviour when code already has exported symbol, eg:
```
use Scalar::Util qw (weaken);
use v5.40;
```
(and vice versa)

leonerd replied:
Symbol visibility is tested by a different test; see the one on line 181.

happy-barney replied:
can you move them closer?

also related test:
```
eval "{ use v5.39 } use v5.20; true";
ok $@, ...
```

leonerd committed:
pad.c: NULL out the pad entry corresponding to a tombstone

When adding a tombstone entry to the pad, there's no need to use a blank
SV for its pad entry, as it will never be used for anything.

This does now mean that pad slots whose names start '&' could now become
NULL, so downstream modules that try to inspect the pad should be a bit
more careful when looking, to see if it is a tombstone. If this commit
proves problematic for CPAN modules, we could undo it and leave all the
pad slots as blank SVt_NULL instead, at a slight increase in memory
usage.

leonerd committed:
Import from `builtin` version bundles should be idempotent on existing symbols

leonerd committed:
Have `use VERSION` and `perl -E` import a corresponding `use builtin` bundle

haarg commented:
Looks good to me.

leonerd commented:
@tonycoz - I rewrote quite a lot of this since your last review. This still good with you?

tonycoz commented on builtin.c:
```diff
@@ -683,8 +683,7 @@ XS(XS_builtin_unimport)
          * tombstone and instead COP_SEQ_MAX_HIGH_set() on the padname to
          * clear it.
          */
-        off = pad_add_name_sv(ampname, padadd_STATE|padadd_TOMBSTONE, 0, 0);
-        SvREFCNT_dec(PL_curpad[off]);
+        pad_add_name_sv(ampname, padadd_STATE|padadd_TOMBSTONE, 0, 0);
```
Does the removal of the SvREFCNT_dec() fix some other problem?

If so the commit message should mention it.

I suspect this was resulting in freeing of a freed scalar further down the line but the commit message should mention it if so.

leonerd replied:
Actually it was indirectly *causing* a problem. Nothing ought to be looking in the pad at the SVs in tombstone entires, but if anything did then previously they'd see the address of an SV that has longsince been freed (or perhaps more likely reused). By not freeing it, we instead leave it as a simple SVt_NULL. I thought that was safer than freeing it then NULL'ing it out, in case something tried to dereference it anyway.

leonerd replied:
I've now made some further adjustments to the code in `pad.c`, and expanded on the wording in the commit message to explain the situation.
